tests/line.c

```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
     SVPNG_END(); /* IEND chunk {} */
 }
 
-#include <math.h>  // ceilf(), floorf(), fminf(), fmaxf(), sinf(), cosf(), sqrtf()
 #include <string.h>  // memset()
+#define Q (20)
+#define Q_PI (1686629713 >> (29 - Q))
+
+/* 32-bit Q format */
+typedef int32_t q_fmt;
+
+/* 64-bit Q format buffer */
+typedef int64_t q_buf;
+
+/* format convertion: float to Q format */
+#define f2Q(x) ((q_fmt) ((x) * (1 << Q)))
+
+/* format convertion: Q format to float */
+#define Q2f(x) (((float) (x)) / (1 << Q))
+
+/* format convertion: Q format to int */
+#define Q2I(x) ((int) ((x) >> Q))
+
+/* format convertion: int to Q format */
+#define I2Q(x) ((q_fmt) ((x) << Q))
+
+/* maximum value of Q format */
+#define QFMT_MAX 0x7FFFFFFF
+
+/* minimum value of Q format */
+#define QFMT_MIN 0x80000000
+
+/* addition of Q format value */
+static inline q_fmt q_add(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a + (q_buf) b;
+    if (tmp > (q_buf) QFMT_MAX)
+        return (q_fmt) QFMT_MAX;
+    if (tmp * -1 >= (q_buf) QFMT_MIN)
+        return (q_fmt) QFMT_MIN;
+    return (q_fmt) tmp;
+};
+
+/* multiplication of Q format value */
+static inline q_fmt q_mul(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a * (q_buf) b;
+
+    /* rounding and resize */
+    tmp += (q_buf) (1 << (Q - 1));
+    tmp >>= Q;
+
+    /* check overflow */
+    if (tmp > (q_buf) QFMT_MAX)
+        return (q_fmt) QFMT_MAX;
+    if (tmp * -1 >= (q_buf) QFMT_MIN)
+        return (q_fmt) QFMT_MIN;
+    return (q_fmt) tmp;
+}
+
+/* division of Q format value */
+static inline q_fmt q_div(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a << Q;
+    if ((tmp >= 0 && b >= 0) || (tmp < 0 && b < 0)) {
+        tmp += (b >> 1);
+    } else {
+        tmp -= (b >> 1);
+    }
+    return (q_fmt) (tmp / b);
+}
+
+/* return the largest integral value that is not greater than x */
+static inline q_fmt floorq(q_fmt x)
+{
+    q_fmt mask = (0xFFFFFFFF >> Q) << Q;
+    return x & mask;
+}
+
+/* return the smallest integral value that is not less than x */
+static inline q_fmt ceilq(q_fmt x)
+{
+    q_fmt mask = (0xFFFFFFFF >> Q) << Q;
+    q_fmt delta = x & ~mask;
+    x = x & mask;
+    return delta ? q_add(x, 1 << Q) : x;
+}
+
+/* return the nonnegative square root of x */
+static inline q_fmt sqrtq(q_fmt x)
```

Comment:
Can you check the following materials accordingly?
* [Fixed Point Square Root](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=71801aa714a149272e58640f2586ddb3a4fa397b)
* [Fixed-Point Arithmetic](https://vanhunteradams.com/FixedPoint/FixedPoint.html)

I would expect the high-level descriptions for square root implementation.

Reply:
I have read the following article, it seems that the fastest algorithm which the article recommended is converting the fixed point value to the floating point value to do 'sqrtf()', and then converting back to the fixed point value result.

Should I change the algorithm into this way? Or just use the original one I made (converging bit by bit)

Reply:
Always measure if something is in doubt.

Reply:
Yes, I have already measured the performance between them in my previous term project:

For perform each algorithm 1000 times and compare their clock cycle:

Convert to floating point and call sqrtf(): 228442
My algorithm: 462650